spec/src/main/asciidoc/configprovider.asciidoc

```diff
@@ -26,23 +26,27 @@ For using Microprofile-Config in a programmatic way the `ConfigProvider` class i
 It allows access to different configurations (represented by a `Config` instance) based on the application in which it is used.
 The `ConfigProvider` internally delegates through to the `ConfigProviderResolver` which contains more low-level functionality.
 
-There are 4 different ways to create an `Config` instance:
+There are 4 different ways to create a `Config` instance:
 
-* In CDI managed components a user can use `@Inject` to access the current application configuration.
+* In CDI managed components, a user can use `@Inject` to access the current application configuration.
   The default and the auto discovered <<configsource,ConfigSources>> will be gathered to form a configuration.
+  The default and the auto discovered <<converters,Converters>> will be gathered to form a configuration.
```

Comment:
Are you sure you meant to use `Converters` here and not `ConfigSources`?  I fail to see how a collection of `Converter` instances constitutes a configuration.

Reply:
OK, the wording for converters should be different. The point is that default and auto discovered converters are used.

Reply:
OK.  Strictly speaking, this isn't true either, is it?  I'm not trying to be difficult: I'm just pointing out that strictly speaking what is required is that conversion happens via `Config#getValue(nameOfInjectionPoint, SomeType.class)`, right?  If so, note that `Converter`s are not in that method's javadoc, or in the specification covering that method's behavior.  (I understand that the specification _implies and hints_ that they _should_ be used _somewhere_ in any given implementation of `Config#getValue(String, Class)`, but it doesn't say that.)  Again, not trying to be annoying! it's just that if this is a spec from which I'm supposed to implement things, I have to connect an awful lot of dots with an awful lot of assumptions right now.

Reply:
I added this to the javadoc of `Config.getValue`:

> Registered converters are applied if they match the propertyType class

I think that it clarifies that converters should be applied. But only those which are registered with the config. For the default config used in injection, default and autodiscovered converters are registered, which is explained here in the spec and also in the javadoc of `ConfigProvider.getConfig()` and `ConfigProperty`.

Is something missing or unclear even after I added these clarifications?

Reply:
I use those words intergangably... Strictly speaking, MicroProfile doesn't certify any implementation, it's rather a recommendation and a set a compatibility tests. There's nothing like impls must do something, or otherwise... But if the word "must" makes things clearer, I can change it

Reply:
I mean, I don't really care.  :-)  It's a question of whether I'm in violation of something if I implement `Config#getValue(String, Class)` without using `Converter`s behind the scenes.  If this is legal, then it's a "should"; if it's not legal (i.e. if there's a TCK test for it or something), then the specification MUST :-) specify it with "must".

Reply:
@ljnelson it is a recommendation on how to create a config. If you write your implementation by completely ignoring the converters by using some reflection per se but you can satisfy the getValue(String, Class) contract, it is absolutely fine. Our TCKs only verify the api behaviours not on how you want to implement it. As @OndrejM pointed out, you can ignore some recommendations implied by the spec, if you can pass the tcks. You will be fine. Therefore, there is no 'must' to force implementors to do the best of practices but the implementors must fulfil the api contracts and pass tcks.

Reply:
Thanks, @Emily-Jiang; now I'm afraid I'm a little more confused.  Is the TCK the source of truth?  For example, the specification does not say anything about whether converters must be used or not.  It describes their behavior assuming they will be used, but there's no requirement that they be used.  But maybe the TCK _does_ enforce their usage (I don't know yet; haven't run it).  Are you saying that in a scenario like this the TCK would be the source of truth: regardless of what the spec says, if my code passes the TCK I'm OK?